src/sctools/bam.py

```diff
@@ -464,29 +500,83 @@ def __repr__(self) -> str:
         return 'query_name'
 
 
-class CellMoleculeGeneQueryNameSortOrder(AlignmentSortOrder):
-    """Hierarchical alignment record sort order (cell barcode >= molecule barcode >= gene name >= query name)."""
+@functools.total_ordering
+class TagSortableRecord(object):
+    """Wrapper for pysam.AlignedSegment that facilitates sorting by tags and query name."""
+
     def __init__(
             self,
-            cell_barcode_tag_key: str = consts.CELL_BARCODE_TAG_KEY,
-            molecule_barcode_tag_key: str = consts.MOLECULE_BARCODE_TAG_KEY,
-            gene_name_tag_key: str = consts.GENE_NAME_TAG_KEY) -> None:
-        assert cell_barcode_tag_key, "Cell barcode tag key can not be None"
-        assert molecule_barcode_tag_key, "Molecule barcode tag key can not be None"
-        assert gene_name_tag_key, "Gene name tag key can not be None"
-        self.cell_barcode_tag_key = cell_barcode_tag_key
-        self.molecule_barcode_tag_key = molecule_barcode_tag_key
-        self.gene_name_tag_key = gene_name_tag_key
-
-    def _get_sort_key(self, alignment: pysam.AlignedSegment) -> Tuple[str, str, str, str]:
-        return (get_tag_or_default(alignment, self.cell_barcode_tag_key, default='N'),
-                get_tag_or_default(alignment, self.molecule_barcode_tag_key, default='N'),
-                get_tag_or_default(alignment, self.gene_name_tag_key, default='N'),
-                alignment.query_name)
+            tag_keys: List[str],
+            tag_values: List[str],
+            query_name: str,
+            record: pysam.AlignedSegment = None):
+        self.tag_keys = tag_keys
+        self.tag_values = tag_values
+        self.query_name = query_name
+        self.record = record
+
+    @classmethod
+    def from_aligned_segment(cls, record: pysam.AlignedSegment, tag_keys: List[str] = None) -> 'TagSortableRecord':
+        """Create a ComparableRecord from a pysam.AlignedSegment and list of tag keys"""
+        assert record is not None
+        if tag_keys is None:
+            tag_keys = []
+        tag_values = []
+        for key in tag_keys:
+            tag_value = get_tag_or_default(record, key, '')
+            tag_values.append(tag_value)
+        query_name = record.query_name
+        return cls(tag_keys, tag_values, query_name, record)
+
+    def __lt__(self, other: 'TagSortableRecord'):
+        self.__verify_tag_keys_match(other)
+        for i in range(len(self.tag_keys)):
+            if self.tag_values[i] < other.tag_values[i]:
+                return True
```

Comment:
Are both of these conditions necessary? This could be condensed into `return self.tag_values[i] < other.tag_values[i]`